buildhub/main/apps.py

```diff
@@ -34,3 +36,15 @@ def _check_s3_bucket_urls():
                 f"settings.SQS_S3_BUCKET_URL ({settings.SQS_S3_BUCKET_URL}) doesn't "
                 "need to be set if it's the same value as settings.S3_BUCKET_URL"
             )
+
+    @staticmethod
+    def _check_bigquery_table():
+        """Check for configured BigQuery credentials and access to the table.
+        On failure, this will raise various exception from the `google.cloud`
+        namespace."""
+        project_id = settings.BQ_PROJECT_ID
+        dataset_id = settings.BQ_DATASET_ID
+        table_id = f"{project_id}.{dataset_id}.{settings.BQ_TABLE_ID}"
+        if settings.BQ_ENABLED:
+            client = bigquery.Client(project=project_id)
+            client.get_table(table_id)
```

Comment:
Hm. When I tried to run locally, this check actually prevents Buildhub from starting up if the table doesn't exist. Maybe we should create the table on startup? We need to deploy the app before we can run the import command, so we do need to deal with this case.

And it looks like it actually prevents the initial import too :laughing::
```
root@b67429ef7a8b:/app# python manage.py rebuild-bigquery
Traceback (most recent call last):
  File "manage.py", line 15, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.7/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.7/site-packages/django/core/management/__init__.py", line 357, in execute
    django.setup()
  File "/usr/local/lib/python3.7/site-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/usr/local/lib/python3.7/site-packages/django/apps/registry.py", line 122, in populate
    app_config.ready()
  File "/app/buildhub/main/apps.py", line 20, in ready
    self._check_bigquery_table()
  File "/app/buildhub/main/apps.py", line 50, in _check_bigquery_table
    client.get_table(table_id)
  File "/usr/local/lib/python3.7/site-packages/google/cloud/bigquery/client.py", line 591, in get_table
    api_response = self._call_api(retry, method="GET", path=table_ref.path)
  File "/usr/local/lib/python3.7/site-packages/google/cloud/bigquery/client.py", line 487, in _call_api
    return call()
  File "/usr/local/lib/python3.7/site-packages/google/api_core/retry.py", line 286, in retry_wrapped_func
    on_error=on_error,
  File "/usr/local/lib/python3.7/site-packages/google/api_core/retry.py", line 184, in retry_target
    return target()
  File "/usr/local/lib/python3.7/site-packages/google/cloud/_http.py", line 421, in api_request
    raise exceptions.from_http_response(response)
google.api_core.exceptions.NotFound: 404 GET https://bigquery.googleapis.com/bigquery/v2/projects/bhearsum-buildhub-bigquery/datasets/buildhubtest/tables/builds2: Not found: Table bhearsum-buildhub-bigquery:buildhubtest.builds2
```

Reply:
I can reproduce this in a test under `test_main_rebuild_bigquery`, which reveals that a missing dataset is also problematic.

```
google.api_core.exceptions.NotFound: 404 POST https://bigquery.googleapis.com/bigquery/v2/projects/amiyaguchi-dev/datasets/testing_not_exists_dataset/tables: Not found: Dataset amiyaguchi-dev:testing_not_exists_dataset
```

The check is working at expected though, since the the table and dataset should exist before insertions can be made. I can add dataset and table creation as part of app startup, though it would make `bigquery:admin` permissions a requirement for the service account.

Reply:
Hm, I thought we were going to require this anyways because of https://github.com/mozilla-releng/buildhub2/pull/617/files#diff-67995dccf929276c3452a5352d57e030R48 ?

Could we pre-provision them before we deploy this, and get rid of the drop/create?

Reply:
That's true, though I'm not entirely sure what roles will be provisioned. If this gets placed under `moz-fx-data-shared-prod`, we will want to make sure that the service account is not able to modify anything outside of its own dataset.

> Could we pre-provision them before we deploy this, and get rid of the drop/create?

In this PR, no. The drop/create clause is there to ensure multiple runs of this command do not create duplicates. We can get rid of this once logic is implemented to only incrementally update the table.

Reply:
Just to be 100% crystal clear, we need credentials and a dataset preprovisioned, and those credentials need to be able to drop and create tables within that dataset?

Reply:
Yes, and those credentials can be tested via the docker-compose tests here.